ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/Linux/procinfo.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+
+ MIT License
+ 
+ Copyright © 2020 Samuel Venable
+ 
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+ 
+ The above copyright notice and this permission notice shall be included in all
+ copies or substantial portions of the Software.
+ 
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ 
+*/
+
+#include <cstdlib>
+#include <cstring>
+
+#include "../procinfo.h"
+
+#include <proc/readproc.h>
+
+using std::string;
+using std::vector;
+using std::to_string;
+
+namespace procinfo {
+
+string path_from_pid(process_t pid) {
+  string path;
+  string link = string("/proc/") + to_string(pid) + string("/exe");
+  char *buffer = realpath(link.c_str(), NULL);
+  path = buffer ? : "";
+  free(buffer);
+  return path;
+}
+
+string cmd_from_pid(process_t pid) {
+  string cmd;
+  proc_t proc_info;
+  if (!pid_exists(pid)) { return ""; }
+  memset(&proc_info, 0, sizeof(proc_info));
+  PROCTAB *proc = openproc(PROC_FILLCOM | PROC_PID, &pid);
+  if (readproc(proc, &proc_info) != 0) {
+    char **args = proc_info.cmdline;
+    size_t cnta = 0;
+    for (; args[cnta] != NULL && strlen(args[cnta]) != 0; cnta++);
+    for (size_t i = 0; i < cnta; i++) {
+      if (string_has_whitespace(args[i])) {
+        cmd += "\"" + string_replace_all(args[i], "\"", "\\\"") + "\"";
+      } else {
+        cmd += args[i];
+      }
+      if (args[i + 1] != NULL) {
+        cmd += " ";
+      }
+    }
+  }

```

Comment:
This environment parsing logic looks similar to BSD, if this is the case, extract a helper for them please. If Win32/Cocoa environment parsing isn't very different either, also extract helpers for them. I am sure the engine is also already doing this elsewhere in ENIGMA too... so eventually this should ALL go to shared/ instead of being duplicated.

Reply:
Thanks for the input, I thought about doing that, but i do feel sometimes adding helper functions for too many aspects of code, especially things that are rather unclear what they actually are supposed to do based on any possible function name, it generally makes code harder to read, but I understand that wanting to not duplicate code is probably more important.

Reply:
That's right, it absolutely is important, the only time to not worry about duplicate code is extremely performance critical stuff and edge cases. This is neither of which and the only thing of concern to us is the long term maintainability of ENIGMA when it has few developers.

Reply:
In other words, since you've said this isn't going in ENIGMA, the decision you must make as a maintain of your own extension Sam is whether you want to minimize errors and maintenance work for yourself, or you care about how quickly the the environment variables can be parsed. The latter isn't much of a concern because the modern world has optimizing compilers that e.g, know how to inline (literally duplicating a function/code by definition) when it should. However, this is actually the same as the static vs dynamic linking debate. The real decision you are making is the classic tradeoff between time and space.